xcdat/bounds.py

```diff
@@ -290,6 +299,97 @@ def add_bounds(self, axis: CFAxisKey) -> xr.Dataset:
+            "freq" or "midpoint".
+
+            * "freq": Create time bounds as the start and end of each
+                timestep's period using either the inferred or specified time
+                frequency (``freq`` parameter). For example, the time bounds
+                will be the start and end of each month for each monthly
+                coordinate point.
+            * "midpoint": Create time bounds using time coordinates as the
+                midpoint between their upper and lower bounds.
+
+        freq : Optional[Literal["year", "month", "day", "hour"]]
+            If ``method="freq"``, this parameter specifies the time frequency
+            for creating time bounds. By default None, which infers the
+            frequency using the time coordinates.
+        daily_subfreq : Literal[1, 2, 3, 4, 6, 8, 12, 24], optional
+            If ``freq=="hour"``, this parameter sets the number of timepoints
+            per day for time bounds, by default None.
+
+            * ``daily_time_freq=None`` infers the daily time frequency from the
+              time coordinates.
+            * ``daily_subfreq=1`` is daily
+            * ``daily_subfreq=2`` is twice daily
+            * ``daily_subfreq=4`` is 6-hourly
+            * ``daily_subfreq=8`` is 3-hourly
+            * ``daily_subfreq=12`` is 2-hourly
+            * ``daily_subfreq=24`` is hourly
+
+        end_of_month : bool, optional
+            If `freq=="month"``, this flag notes that the timepoint is saved
+            at the end of the monthly interval (see Note), by default False.
+            Some timepoints are saved at the end of the interval, e.g., Feb. 1
+            00:00 for the time interval Jan. 1 00:00 - Feb. 1 00:00. Since this
+            method determines the month and year from the time vector, the
+            bounds will be set incorrectly if the timepoint is set to the end of
+            the time interval. For these cases, set ``end_of_month=True``.
+
+        Returns
+        -------
+        xr.Dataset
+            The dataset with bounds added.
+
+        """
+        ds = self._dataset.copy()
+        coord_vars: Union[xr.DataArray, xr.Dataset] = get_dim_coords(self._dataset, "T")
+
+        for coord in coord_vars.coords.values():
+            # Check if the coord var has a "bounds" attr and the bounds actually
+            # exist in the Dataset. If it does not, then add the bounds.
+            try:
+                bounds_key = ds[coord.name].attrs["bounds"]
+                ds[bounds_key]
+
+                continue
+            except KeyError:
+                if method == "freq":
+                    bounds = self._create_time_bounds(
+                        coord, freq, daily_subfreq, end_of_month
+                    )
+                elif method == "midpoint":
+                    bounds = self._create_bounds("T", coord)
+
+                ds[bounds.name] = bounds
+                ds[coord.name].attrs["bounds"] = bounds.name
```

Comment:
`add_missing_bounds()` has these try and except statements because we want to make sure datasets can be opened with `open_dataset()` and `open_mfdataset()`.

`add_bounds()` and `add_time_bounds()` are standalone public APIs that are called on datasets that were successfully opened. I lean towards keeping these try and except statements, otherwise those errors will be silent. The user has more control over how to use these APIs and can add a try and except statement in their own code.

```python
ds = xc.open_mfdataset(...)

try:
  ds = ds.add_bounds(axis="X")
except (ValueError, TypeError):
  continue

try:
  ds = ds.add_time_bounds()
except (ValueError, TypeError):
  continue
```

Reply:
Is `height` a time axis? It seems like the user shouldn't get a ValueError trying to add bounds for `time` when there is a non-dimensional height coordinate embedded in the data.

Reply:
SIngleton coordinate variables are included alongside other coordinate variables. This issue is related to https://github.com/xCDAT/xcdat/issues/389.

I'll investigate a solution similar to https://github.com/xCDAT/xcdat/pull/392/files.